state-transition/core/state_processor_forks.go

```diff
@@ -0,0 +1,190 @@
+// SPDX-License-Identifier: BUSL-1.1
+//
+// Copyright (C) 2025, Berachain Foundation. All rights reserved.
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file of this repository and at www.mariadb.com/bsl11.
+//
+// ANY USE OF THE LICENSED WORK IN VIOLATION OF THIS LICENSE WILL AUTOMATICALLY
+// TERMINATE YOUR RIGHTS UNDER THIS LICENSE FOR THE CURRENT AND ALL OTHER
+// VERSIONS OF THE LICENSED WORK.
+//
+// THIS LICENSE DOES NOT GRANT YOU ANY RIGHT IN ANY TRADEMARK OR LOGO OF
+// LICENSOR OR ITS AFFILIATES (PROVIDED THAT YOU MAY USE A TRADEMARK OR LOGO OF
+// LICENSOR AS EXPRESSLY REQUIRED BY THIS LICENSE).
+//
+// TO THE EXTENT PERMITTED BY APPLICABLE LAW, THE LICENSED WORK IS PROVIDED ON
+// AN “AS IS” BASIS. LICENSOR HEREBY DISCLAIMS ALL WARRANTIES AND CONDITIONS,
+// EXPRESS OR IMPLIED, INCLUDING (WITHOUT LIMITATION) WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT, AND
+// TITLE.
+
+package core
+
+import (
+	"fmt"
+
+	"github.com/berachain/beacon-kit/consensus-types/types"
+	"github.com/berachain/beacon-kit/primitives/common"
+	"github.com/berachain/beacon-kit/primitives/constants"
+	"github.com/berachain/beacon-kit/primitives/math"
+	"github.com/berachain/beacon-kit/primitives/version"
+	statedb "github.com/berachain/beacon-kit/state-transition/core/state"
+)
+
+// PrepareStateForFork prepares the state for the fork version at the given timestamp.
+//   - If this function is called for the same version as the state's current version,
+//     it will do nothing. Unless it is the genesis slot, in which case we want to
+//     prepare the state for the genesis fork version.
+//   - If this function is called for a version before the state's current version,
+//     it will return error as this is not allowed.
+//   - If this function is called for a version after the state's current version,
+//     it will upgrade the state to the new version.
+func (sp *StateProcessor) PrepareStateForFork(
+	st *statedb.StateDB, timestamp math.U64, slot math.Slot, logUpgrade bool,
```

Comment:
IIUC, slot here is always the `st.GetSlot`. Even in case where we are building a block, state would have already be updated to the right slot via ProcessSlot.
Consider dropping slot from the parameters and load it from `st`. This is what we do across all state processing functions

Reply:
https://github.com/berachain/beacon-kit/pull/2678/commits/6408e427da6c39cd3f83aa345b62a03ed030ad78